homeassistant/components/homekit/type_air_purifiers.py

```diff
@@ -214,202 +254,131 @@ def run(self) -> None:
 
         Run inside the Home Assistant event loop.
         """
-        if self.linked_humidity_sensor:
-            self._subscriptions.append(
-                async_track_state_change_event(
-                    self.hass,
-                    [self.linked_humidity_sensor],
-                    self._async_update_current_humidity_event,
-                    job_type=HassJobType.Callback,
-                )
-            )
-
-        if self.linked_pm25_sensor:
-            self._subscriptions.append(
-                async_track_state_change_event(
-                    self.hass,
-                    [self.linked_pm25_sensor],
-                    self._async_update_current_pm25_event,
-                    job_type=HassJobType.Callback,
-                )
-            )
-
-        if self.linked_temperature_sensor:
-            self._subscriptions.append(
-                async_track_state_change_event(
-                    self.hass,
-                    [self.linked_temperature_sensor],
-                    self._async_update_current_temperature_event,
-                    job_type=HassJobType.Callback,
+        # Add subscriptions for all float sensors
+        float_sensors: list[
+            tuple[
+                str | None,
+                Characteristic | None,
+                Callable[[float], None] | None,
+            ]
+        ] = [
+            (
+                self.linked_humidity_sensor,
+                self.char_current_humidity,
+                None,
+            ),
+            (
+                self.linked_pm25_sensor,
+                self.char_pm25_density,
+                self._async_update_air_quality_from_pm25,
+            ),
+            (
+                self.linked_temperature_sensor,
+                self.char_current_temperature,
+                None,
+            ),
+            (
+                self.linked_filter_life_level_sensor,
+                self.char_filter_life_level,
+                self._async_update_filter_change_indicator_from_filter_life_level,
+            ),
+        ]
+        for linked_sensor, char, post_processor in float_sensors:
+            if linked_sensor and char:
+                self._subscriptions.append(
+                    async_track_state_change_event(
+                        self.hass,
+                        [linked_sensor],
+                        partial(  # type: ignore[misc] # mypy fails to infer the type parameter (`float` here)
+                            self._async_update_sensor_event,
+                            char,
+                            linked_sensor,
+                            convert_to_float,
+                            post_processor=post_processor,
+                        ),
+                        job_type=HassJobType.Callback,
+                    )
                 )
-            )
```

Comment:
Since we have the `async_track_state_change_event` pattern in a lot more accessory types as well, we could add an new helper method in the base `HomeAccessory` class for tracking linked sensor that would make the `async_track_state_change_event` calls and take care getting the data in the right format. Future PRs could than update the other accessory types

Reply:
Good idea, just pushed a few commits for that.